node/subsystem-util/src/database.rs

```diff
@@ -179,58 +175,47 @@ pub mod paritydb_impl {
 			map_err(self.db.get(col as u8, key))
 		}
 
-		fn get_by_prefix(&self, col: u32, prefix: &[u8]) -> Option<Box<[u8]>> {
-			self.iter_with_prefix(col, prefix).next().map(|(_, v)| v)
+		fn get_by_prefix(&self, col: u32, prefix: &[u8]) -> Result<Option<DBValue>> {
+			self.iter_with_prefix(col, prefix)
+				.next()
+				.transpose()
+				.map(|mb| mb.map(|(_, v)| v))
 		}
 
-		fn iter<'a>(&'a self, col: u32) -> Box<dyn Iterator<Item = (Box<[u8]>, Box<[u8]>)> + 'a> {
-			let mut iter = handle_err(self.db.iter(col as u8));
+		fn iter<'a>(&'a self, col: u32) -> Box<dyn Iterator<Item = Result<DBKeyValue>> + 'a> {
+			let mut iter = match self.db.iter(col as u8) {
+				Ok(iter) => iter,
+				Err(e) => return Box::new(std::iter::once(map_err(Err(e)))),
+			};
 			Box::new(std::iter::from_fn(move || {
-				if let Some((key, value)) = handle_err(iter.next()) {
-					Some((key.into_boxed_slice(), value.into_boxed_slice()))
-				} else {
-					None
-				}
+				iter.next().transpose().map(|r| map_err(r.map(|(k, v)| (k.into(), v))))
 			}))
 		}
 
 		fn iter_with_prefix<'a>(
 			&'a self,
 			col: u32,
 			prefix: &'a [u8],
-		) -> Box<dyn Iterator<Item = (Box<[u8]>, Box<[u8]>)> + 'a> {
+		) -> Box<dyn Iterator<Item = Result<DBKeyValue>> + 'a> {
 			if prefix.len() == 0 {
 				return self.iter(col)
 			}
-			let mut iter = handle_err(self.db.iter(col as u8));
```

Comment:
@cheme I haven't removed handle_err in `write`, but it would be nice to do eventually.

Reply:
Oh yes, I missed it in the kvdb PR, next time.

Reply:
I meant to say that part of #4939 is now possible/easier to implement. But I'm leaving it out of scope of this PR.